File: kotlin-swift-interopedia-samples-skie/shared/src/commonMain/kotlin/com/jetbrains/swiftinteropplayground/functions/DSL.kt
package com.jetbrains.swiftinteropplayground.functions

@DslMarker
annotation class DslMarkerExample

data class Experiment(val key: String, val description: String)

@DslMarkerExample
data class ExperimentsDsl(val list: MutableList<Experiment> = mutableListOf<Experiment>()) {
    fun enable(experiment: Experiment) {
        list += experiment
    }

}

@DslMarkerExample
data class Dsl(val experiments: ExperimentsDsl = ExperimentsDsl()) {
    fun experiments(block: ExperimentsDsl.() -> Unit = {}) {
        experiments.block()
    }
}

fun example() {
    val dsl = Dsl().apply {
        experiments {
            enable(Experiment(key = "key1", description = "desc1"))
        }
    }
    println(dsl)
}

File: kotlin-swift-interopedia-samples-without-skie/iosApp/iosApp/Functions/DSL.swift
import Foundation
import shared

class DslExample{
    
    private func dslBlock(block: (Dsl) -> Dsl) -> Dsl {
        var dsl = Dsl()
        dsl = block(dsl)
        return dsl
    }
    
    func example(){
        dslBlock { dsl in
            dsl.experiments { e in e.enable(experiment: Experiment(key: "key1", description: "param123"))}
            return dsl
        }
    }
}
// var l1 = Dsl(e: ExperimentsDsl(l: NSMutableArray()))
//l1.experiments{ a in a.enable(experiment: Experiment(key: "key1", description: "param1"))
//}
//print("DSL is \(l1)")
